code/go/0chain.net/core/common/rate_limiter.go

```diff
@@ -1,159 +1,89 @@
-func UseUserRateLimit(h http.Handler) http.Handler {
-	if !userRateLimit.RateLimit {
-		return h
+	if ipLookups != nil {
+		lmt.SetIPLookups(ipLookups)
 	}
 
-	return tollbooth.LimitFuncHandler(userRateLimit.Limiter, h.ServeHTTP)
+	lmt.SetIgnoreURL(ignoreUrl)
+	return lmt
 }
 
-// FileRateLimit is a custom of `tollbooth.LimitHandlerFunc` that uses custom values to build keys for limiting file actions.
-func FileRateLimit(handler ReqRespHandlerf) ReqRespHandlerf {
-	if fileRateLimit == nil || !fileRateLimit.RateLimit {
-		return handler
-	}
-
-	lmt := fileRateLimit.Limiter
-
+func RateLimit(handler ReqRespHandlerf, lmt *limiter.Limiter) ReqRespHandlerf {
 	return func(w http.ResponseWriter, r *http.Request) {
-		keys := FileLimitBuildKeys(lmt, r)
-		httpError := tollbooth.LimitByKeys(lmt, keys)
-		if httpError != nil {
-			lmt.ExecOnLimitReached(w, r)
-			if lmt.GetOverrideDefaultResponseWriter() {
-				return
+		TryParseForm(r)
+
+		if lmt != nil {
+
+			keys := tollbooth.BuildKeys(lmt, r)
+			clientID := r.Header.Get(ClientHeader)
+
+			keys = append(keys, []string{clientID})
+
+			for _, k := range keys {
+				httpError := tollbooth.LimitByKeys(lmt, k)
+				if httpError != nil {
+					logging.Logger.Error(fmt.Sprintf("Rate limit error: %s", httpError.Error()))
+					lmt.ExecOnLimitReached(w, r)
+					setResponseHeaders(lmt, w, r)
+					w.Header().Add("Content-Type", lmt.GetMessageContentType())
+					w.WriteHeader(httpError.StatusCode)
+					w.Write([]byte(httpError.Message)) // nolint
+					return
+				}
 			}
-			w.Header().Add("Content-Type", lmt.GetMessageContentType())
-			w.WriteHeader(httpError.StatusCode)
-			_, _ = w.Write([]byte(httpError.Message))
-			return
 		}
-
-		// There's no rate-limit error, serve the next handler.
 		handler(w, r)
 	}
 }
 
-// FileLimitBuildKeys return build keys to be used for file rate limiter.
-func FileLimitBuildKeys(lmt *limiter.Limiter, r *http.Request) []string {
-	baseKeys := tollbooth.BuildKeys(lmt, r)
-
-	client := r.Header.Get(ClientHeader)
-	baseKeys[0] = append(baseKeys[0], strings.ToLower(client))
-
-	// add connection_id when present (exists for upload)
-	conn, ok := requestField(r, "connection_id")
-	if ok {
-		baseKeys[0] = append(baseKeys[0], strings.ToLower(conn))
-		return baseKeys[0]
-	}
-
-	// OR add path hash (may exists for download)
-	pathHash := r.Header.Get("X-Path-Hash")
-	if pathHash != "" {
-		baseKeys[0] = append(baseKeys[0], strings.ToLower(pathHash))
-		return baseKeys[0]
-	}
-
-	// OR add path (may exists for download)
-	path := r.Header.Get("X-Path")
-	if path != "" {
-		baseKeys[0] = append(baseKeys[0], path)
-		return baseKeys[0]
+func RateLimitByIP(handler ReqRespHandlerf, lmt *limiter.Limiter) ReqRespHandlerf {
+	return func(w http.ResponseWriter, r *http.Request) {
+		TryParseForm(r)
+
+		if lmt != nil {
+
+			keys := tollbooth.BuildKeys(lmt, r)
+			for _, k := range keys {
+				httpError := tollbooth.LimitByKeys(lmt, k)
```

Comment:
why do we need to limit for every key here, it is all headers is it really needed?

Reply:
Keys is of type `[][]string`.
`tollbooth.BuildKeys(lmt, r)` will return `[[keys extracted from request headers...]]` and its length is 1. So it is single iteration only.

It is equivalent of doing, `httpError := tollbooth.LimitByKeys(lmt, keys[0])`

Reply:
I can't get what we are doing here, we are computing key of the bucket by concatenating all the headers, what is it for, should we use only one header here, like ip or forwarded-ip?